Add color and typography presets to Global Styles

scruffian opened the pull request:
## What?
A popular feature in classic themes was to allow users to change the color palette and font families of their theme, usually the customizer. This is already possible in Global Styles, but the UI to do so is tricky to use - it takes a lot of clicks and a deep understanding of how Global Styles works. This PR abstracts that complexity away for users so they can change colors and typography details easily without knowing how Global Styles works.

These color and typography settings are extracted from the style variations, and exposed in the Global Styles UI. 

## Why?
This makes it possible to mix and match color and typography settings from different style variations. In the future this will also allow theme developers to offer a much wider range of color and typography variations and potentially even mix and match these variations between themes.

## How?
- Process style variations to extract the color or typography rules from them. Note: This extracts all of the typography/color rules, not just those under "settings", but all of the rules in the style variation.
- Add a new UI to the colors and typography sections of the Global Styles UI to give access to these extracted settings

## Testing Instructions
1. Switch to a theme that has multiple style variations
2. Open the Site Editor
3. Go to the Global Styles panel
4. Open the Colors section
5. Open the Palette section
6. You should see a section called "Presets"
7. Select one of the presets
8. The colors of your preview should be updated to match the preset
9. Go back to the root of Global Styles
10. Open the Typography section
11. Select "Typesets"
12. Select one of the Preset typography options
13. The typography of the preview should update

## Screenshots or screencast <!-- if applicable -->
<img width="280" alt="Screenshot 2023-12-07 at 17 07 08" src="https://github.com/WordPress/gutenberg/assets/275961/86a1c814-f860-4d3a-bb70-ad94464262b3">
<img width="280" alt="Screenshot 2023-12-07 at 16 48 04" src="https://github.com/WordPress/gutenberg/assets/275961/ca559b7c-efe4-4c58-9a8d-2442880dc033">
<img width="281" alt="Screenshot 2023-12-07 at 16 47 52" src="https://github.com/WordPress/gutenberg/assets/275961/9c3c0086-334e-4d32-aae4-3c0952a9cfe9">

There's also a demo video here: https://ben.blog/2023/12/07/global-styles-add-color-and-typography-presets/
## Todo
- Show the defaults for both color and typography
- Also apply the duotone from the style variation

## Questions
- Should we name each color/typography variation? What should they be called?
- Does the "Typesets" section in "Typography" make sense?
- How could we surface this in Site View?


## Notes
One of the powerful things about the approach taken here is that it will work for existing themes - it doesn't require theme developers to add special settings to their themes.

scruffian committed:
Add colors and typography to the styles tab

scruffian committed:
refactor

scruffian committed:
refactor

scruffian committed:
Add font family names

scruffian committed:
add heading and body previews

scruffian committed:
use palettes

scruffian committed:
make typograpgy a grid

scruffian committed:
make the styles look nicer

scruffian committed:
move component to its own file

scruffian committed:
move color and type variations to different locations in the UI

scruffian committed:
revert

scruffian committed:
revert

scruffian committed:
revert

scruffian committed:
revert

scruffian committed:
code dedupe

scruffian committed:
Add typesets

scruffian committed:
reorg typogrpahy

scruffian committed:
tidy up

scruffian committed:
tidy up code

scruffian committed:
restyle color palettes

scruffian committed:
restyle color palettes

scruffian committed:
remove typesets

scruffian committed:
add padding back

scruffian commented on packages/edit-site/src/components/global-styles/screen-typography-typesets.js:
```diff
@@ -0,0 +1,104 @@
+
+const { GlobalStylesContext } = unlock( blockEditorPrivateApis );
+
+const getFontFamilies = ( themeJson ) => {
```
There's a lot of duplicate code in this function that could be extracted to a function.

scruffian commented on packages/edit-site/src/components/global-styles/screen-typography-typesets.js:
```diff
@@ -0,0 +1,104 @@
+};
+
+export default function ScreenTypographyTypesets() {
+	const variations = useSelect( ( select ) => {
```
Return an object.

scruffian commented on packages/edit-site/src/components/global-styles/screen-typography-typesets.js:
```diff
@@ -0,0 +1,104 @@
+	const { base, user } = useContext( GlobalStylesContext );
+
+	const typographyVariations =
+		variations && getVariationsByProperty( user, variations, 'typography' );
```
Add the default?

scruffian committed:
refactor

scruffian commented on packages/edit-site/src/components/global-styles/typeset.js:
```diff
@@ -0,0 +1,103 @@
+/**
```
This file isn't used in the end.

ramonjd replied:
I removed it - hope that's okay.

scruffian commented on packages/edit-site/src/components/global-styles/utils.js:
```diff
@@ -10,3 +15,80 @@ export function getVariationClassName( variation ) {
+ * @param {Object} object
+ * @return {Object} The cloned object.
+ */
+function cloneDeep( object ) {
```
Most of these functions are generic enough that they should live in a more general library IMO.

ramonjd replied:
Also used here in the edit-site package here:

https://github.com/WordPress/gutenberg/blob/cde700936c358f6e99d926ca336464c4b6b098b2/packages/edit-site/src/hooks/push-changes-to-global-styles/index.js#L278-L278

scruffian commented on packages/edit-site/src/components/global-styles/variations-color.js:
```diff
@@ -0,0 +1,152 @@
+		} );
+	};
+
+	const selectOnEnter = ( event ) => {
```
Quite a lot of this is just copy/paste from the Variation component.

scruffian commented on packages/edit-site/src/components/global-styles/variations-color.js:
```diff
@@ -0,0 +1,152 @@
+
+export default function ColorVariations() {
+	const { user } = useContext( GlobalStylesContext );
+	const variations = useSelect( ( select ) => {
```
Return an object

scruffian commented on packages/edit-site/src/components/global-styles/variations-color.js:
```diff
@@ -0,0 +1,152 @@
+		).__experimentalGetCurrentThemeGlobalStylesVariations();
+	}, [] );
+	const colorVariations =
+		variations && getVariationsByProperty( user, variations, 'color' ); // should also get filter?
```
Add the default variation

scruffian commented on packages/edit-site/src/components/global-styles/variations-typography.js:
```diff
@@ -0,0 +1,219 @@
+}
+
+export default function TypographyVariations() {
+	const variations = useSelect( ( select ) => {
```
Return an object

scruffian commented on packages/edit-site/src/components/global-styles/variations-typography.js:
```diff
@@ -0,0 +1,219 @@
+
+	const { base, user } = useContext( GlobalStylesContext );
+
+	const typographyVariations =
```
Add the default variation

MaggieCabrera commented:
This doesn't seem to be working for variations with more than 2 fonts. I added this variation to TT4 that uses a third font for buttons and it didn't show under typesets:

```
{
	"$schema": "https://schemas.wp.org/trunk/theme.json",
	"version": 2,
	"title": "Maggie",
	"settings": {
		"typography": {
			"fluid": true,
			"fontFamilies": [
				{
					"fontFace": [
						{
							"fontFamily": "Instrument Sans",
							"fontStyle": "normal",
							"fontWeight": "400 700",
							"src": [
								"file:./assets/fonts/instrument-sans/InstrumentSans-VariableFont_wdth,wght.woff2"
							]
						},
						{
							"fontFamily": "Instrument Sans",
							"fontStyle": "italic",
							"fontWeight": "400 700",
							"src": [
								"file:./assets/fonts/instrument-sans/InstrumentSans-Italic-VariableFont_wdth,wght.woff2"
							]
						}
					],
					"fontFamily": "\"Instrument Sans\", sans-serif",
					"name": "Instrument Sans",
					"slug": "heading"
				},
				{
					"fontFace": [
						{
							"fontFamily": "Cardo",
							"fontStyle": "normal",
							"fontWeight": "400",
							"src": ["file:./assets/fonts/cardo/cardo_normal_400.woff2"]
						},
						{
							"fontFamily": "Cardo",
							"fontStyle": "italic",
							"fontWeight": "400",
							"src": ["file:./assets/fonts/cardo/cardo_italic_400.woff2"]
						},
						{
							"fontFamily": "Cardo",
							"fontStyle": "normal",
							"fontWeight": "700",
							"src": ["file:./assets/fonts/cardo/cardo_normal_700.woff2"]
						}
					],
					"fontFamily": "Cardo",
					"name": "Cardo",
					"slug": "buttons"
				},
				{
					"fontFace": [
						{
							"fontFamily": "Jost",
							"fontStyle": "normal",
							"fontWeight": "100 900",
							"src": ["file:./assets/fonts/jost/Jost-VariableFont_wght.woff2"]
						},
						{
							"fontFamily": "Jost",
							"fontStyle": "italic",
							"fontWeight": "100 900",
							"src": [
								"file:./assets/fonts/jost/Jost-Italic-VariableFont_wght.woff2"
							]
						}
					],
					"fontFamily": "\"Jost\", sans-serif",
					"name": "Jost",
					"slug": "body"
				},
				{
					"fontFamily": "-apple-system, BlinkMacSystemFont, avenir next, avenir, segoe ui, helvetica neue, helvetica, Cantarell, Ubuntu, roboto, noto, arial, sans-serif",
					"name": "System Sans-serif",
					"slug": "system-sans-serif"
				},
				{
					"fontFamily": "Iowan Old Style, Apple Garamond, Baskerville, Times New Roman, Droid Serif, Times, Source Serif Pro, serif, Apple Color Emoji, Segoe UI Emoji, Segoe UI Symbol",
					"name": "System Serif",
					"slug": "system-serif"
				}
			]
		}
	},
	"styles": {
		"blocks": {
			"core/button": {
				"typography": {
					"fontFamily": "var(--wp--preset--font-family--buttons)"
				}
			}
		},
		"elements": {
			"button": {
				"typography": {
					"fontFamily": "var(--wp--preset--font-family--buttons)"
				}
			}
		}
	}
}

```

ramonjd commented:
@scruffian @MaggieCabrera If you need help moving this one forward, I'm happy to rebase and test, and tinker with it. 

I'll take it for a spin next week. 

If there are any outstanding items or gotchas, let me know. 

Thank you!

ramonjd commented on packages/edit-site/src/components/global-styles/variations-typography.js:
```diff
@@ -0,0 +1,219 @@
+
+	typographyVariations?.forEach( ( variation ) => {
+		const [ bodyFontFamilyName, headingFontFamilyName ] =
+			getFontFamilyNames( mergeBaseAndUserConfigs( base, variation ) );
```
Maybe we need to come up with a way to name style variations with no font-families, e.g., cases where variations have typography styles only like this:

```json
{
	"settings": {},
	"styles": {
		"typography": {
			"letterSpacing": "10px"
		},
		"blocks": {
			"core/post-title": {
				"typography": {
					"fontWeight": "700"
				},
				"color": {
					"background": "red"
				}
			}
		}
	}
}
```

Possibly `${variationTitle} - typeset` or, when `variation.title` doesn't exist in the theme.json `${variationFileName} - typeset`?

<img width="278" alt="Screenshot 2024-02-05 at 2 17 05 pm" src="https://github.com/WordPress/gutenberg/assets/6458278/fa02c3b9-729d-4728-b053-69d67d7ab997">

richtabor replied:
> Maybe we need to come up with a way to name style variations with no font-families, e.g., cases where variations have typography styles only like this:

To simplify this, I propose we lean in on a "Font book" type approach, where the heading and body fonts are used in for `A` and `a` respectively, like this: 

![CleanShot 2024-02-06 at 13 36 34](https://github.com/WordPress/gutenberg/assets/1813435/460ad2b2-3a37-4740-a8e5-f5fb76c126a9)

We already do that for styles: 

![CleanShot 2024-02-06 at 13 38 10](https://github.com/WordPress/gutenberg/assets/1813435/ede48aba-8233-4375-8032-23c3ab3edb86)

scruffian replied:
I updated the PR to use these designs:

<img width="315" alt="Screenshot 2024-02-24 at 10 34 34" src="https://github.com/WordPress/gutenberg/assets/275961/5067db96-8bd8-4862-82c3-1e2c3da100c2">

mtias replied:
Yeah, this looks good. We can expand later on listing exactly what aspects are being modified, which is a general concern of variations.

ramonjd replied:
Looks great!

ramonjd commented on packages/edit-site/src/components/global-styles/screen-typography.js:
```diff
@@ -23,7 +31,28 @@ function ScreenTypography() {
+						<VStack spacing={ 3 }>
+							<FontFamilies />
+							<ItemGroup isBordered>
+								<NavigationButtonAsItem
```
Wondering if the all the work grabbing the typography settings from the variations could be done in a memoized hook, or store selector. 

That way we could do a length check on the array here and skip rendering the button.

ramonjd commented on packages/edit-site/src/components/global-styles/variations-color.js:
```diff
@@ -0,0 +1,152 @@
+		);
+	}
+
+	const colors = variation?.settings?.color?.palette?.theme ?? [];
```
Same with typography, a variation might have color values in the styles. 

<img width="268" alt="Screenshot 2024-02-05 at 2 44 44 pm" src="https://github.com/WordPress/gutenberg/assets/6458278/7d3e7fd9-b4fe-488c-9154-756385248c01">


Should we skip rendering the palette if there are no color presets? 

Or maybe an icon instead of the `<ColorIndicator />`

<img width="267" alt="Screenshot 2024-02-05 at 2 54 42 pm" src="https://github.com/WordPress/gutenberg/assets/6458278/3955ffa3-bf54-472a-a8dd-c0c7c179f817">

richtabor replied:
> Should we skip rendering the palette if there are no color presets?

Yes. I wouldnt expect a color palette preset to be available, if there are no colors defined.

ramonjd replied:
No probs. 

It raises another question: should pressing the theme style variation color preset button also apply all typography related styles to the editor (inc. element and block typography styles) or just the typography presets?

The current state of this PR is that it applies all typography styles and settings in the theme.json tree.

scruffian replied:
> The current state of this PR is that it applies all typography styles and settings in the theme.json tree.

I believe this is the correct behaviour.

ramonjd committed:
Show theme_name + style where there is no typography font family info.
Remove unused file.

ramonjd reviewed (commented):
Just rebasing/testing today and making some observations. I can look further into the following:

- create hooks to grab typography/color variation sets
- decide how to handle labelling variations with no presets
- tests
- other stuff 😄

ramonjd committed:
Consolidating logic into a hook phase 1

ramonjd committed:
2022 borks getFamilyPreviewStyle because `face.fontStyle` doesn't exist

ramonjd committed:
comments

ramonjd commented:
I'm thinking we could split this PR. 

First we could work on the hook to grab theme style variation styles by property:

- https://github.com/WordPress/gutenberg/pull/56622/files#diff-7b473a2598fb33ebeacdb9e4ad6a621ebe67e96938744ad11cbf25f08c66d477

This could unblock the @apeatling more quickly.

This would mean we'd work on the UI/site editor components etc in another PR. There are some aspects of how and what we show in the editor that I think need resolved, for instance how should the site editor, if at all, present theme style variations that have no presets? Related comments: https://github.com/WordPress/gutenberg/pull/56622/files#r1477640893 and https://github.com/WordPress/gutenberg/pull/56622/files#r1477627148 Maybe we shouldn't focus on presets so much and rather present them as subsets of variations (typography and so on) So if a theme hasn't got a any heading/body font families defined, what do we do?

In short, in my estimation, the discussions about UX and UI will take longer than getting some working utils ready.

What do folks think?

richtabor commented:
> Maybe we shouldn't focus on presets so much and rather present them as subsets of variations (typography and so on) 

I think they're one and the same. Having a full theme style variation is a nice way to switch up your whole site's look, but splitting out the colors and typesets within each is nice so you're not having to go all-or-nothing. 

> So if a theme hasn't any heading/body font families defined, what do we do?

I wouldn't expect there to be a preset for it then.

ramonjd commented:
Thanks for the feedback @richtabor! This gives us plenty to go off, thanks!

scruffian commented:
> In short, in my estimation, the discussions about UX and UI will take longer than getting some working utils ready.

I agree with this. There's a whole lot of crazy javascript manipulations happening that need to be tidied up and tested thoroughly. We can get a lot of this in place before making the UX decisions.

Thanks for working on this.

ramonjd commented on packages/edit-site/src/components/global-styles/font-library-modal/utils/preview-styles.js:
```diff
@@ -60,7 +60,8 @@ export function getFamilyPreviewStyle( family ) {
 		const normalFaces = family.fontFace.filter(
-			( face ) => face.fontStyle.toLowerCase() === 'normal'
+			( face ) =>
+				face?.fontStyle && face.fontStyle.toLowerCase() === 'normal'
```
This change was to prevent fatals as we're passing the function style variations without presets.

ramonjd committed:
Filtering results

ramonjd commented:
WIP utils PR here:

- https://github.com/WordPress/gutenberg/pull/58803

ramonjd committed:
Create new convenience hook `useCurrentMergeThemeStyleVariationsWithUserConfig()`
Replace usage with utils created in https://github.com/WordPress/gutenberg/pull/58803

ramonjd committed:
add a todo

ramonjd commented on packages/edit-site/src/components/global-styles/color-palette-panel.js:
```diff
@@ -89,6 +96,10 @@ export default function ColorPalettePanel( { name } ) {
 			/>
+			{ /* @TODO: pass down variations to component? */ }
+			{ !! colorVariations.length && (
+				<ColorVariations variations={ colorVariations } />
```
I'm playing around and wondering... for colors, selecting a variation merges all color values into the current user styles and therefore the editor canvas: settings, styles, elements, blocks and so on. 

It doesn't just replace current presets/palettes.

Since it's under "Palettes" at the moment, should it just replace color settings, and therefore presets?

This would be consistent behaviour, for this panel at least.

Otherwise, if we're merging all a variation's  colors to the current user styles, maybe this component be moved from under "palettes" up a level to the colors, and renamed... to, I don't know, something like "Apply a color variation" or something.

scruffian replied:
I think it should take all the color settings, but I'll defer to @richtabor

richtabor replied:
I think it's fine to leave the presets in Colors > Palette, as they are presets for the palette first (which consequently change the other values, as expected). Otherwise there's too much of a disconnect between the presets and the palette. 

I'd consider moving it above "Custom" even, as those custom colors are appended to the palette. 

![CleanShot 2024-02-16 at 11 28 47](https://github.com/WordPress/gutenberg/assets/1813435/baf0541d-a9ee-426f-b08c-4abf27a25f21)

ramonjd committed:
A bit of clean up. Remove comments, adjust margins.

ramonjd commented:
I think this PR is ready for post-experimental review 😄 

I'm not sure about placement or copy, but the basic functionality is there.

There's a bit of repetition around variation components, but I think it's something that can be cleaned up in a follow up. 

> Should we name each color/typography variation? What should they be called?

My vote would go to the style variation title, if available. If it's not, perhaps the file name minus `.json` extension?

> Does the "Typesets" section in "Typography" make sense?

Typeset, I think, is mainly used as a verb to set words in type for publishing. Maybe it's okay, but it might cause some confusion. Typography sets, or Typography styles might better reflect the behaviour... 🤷🏻 

But since we're using "Presets" for colors, why not here as well?

I'm thinking "preset" in the everyday sense, not theme.json specific, e.g., settings prepared or set for a user beforehand.

> How could we surface this in Site View?

"Choose different style combinations" is a feature of this issue:

- https://github.com/WordPress/gutenberg/issues/53483#issuecomment-1951325374

Could we take inspiration from there?

scruffian committed:
move variation to a new component

scruffian committed:
make variation a composible component

scruffian commented on packages/edit-site/src/components/global-styles/variations-typography.js:
```diff
@@ -0,0 +1,221 @@
+	return [ bodyFontFamily?.name, headingFontFamily?.name ];
+};
+
+function TypographyVariation( { variation } ) {
```
TODO - use the variation component here.

scruffian replied:
Done in cf8322f054

scruffian committed:
use the variation component for typography

scruffian committed:
Update typography preview

scruffian committed:
Add animation

scruffian committed:
move variation to a component

scruffian committed:
move preset to the higher level

scruffian committed:
Remove unneeded CSS

scruffian committed:
updawte the design

scruffian commented on packages/edit-site/src/components/global-styles/preview-colors.js:
```diff
@@ -0,0 +1,208 @@
+/**
```
This is a duplicate of the preview component

scruffian commented:
I think this is in a pretty good place. The main thing left to do is to deal with the duplicate code in `packages/edit-site/src/components/global-styles/preview-colors.js`, although that could potentially be a followup.

scruffian committed:
remove unsed css

mtias commented on packages/edit-site/src/components/global-styles/style.scss:
```diff
@@ -199,3 +204,9 @@
 	fill: currentColor;
 }
+
+.edit-site-global-styles-type-preview {
```
are these styles something that can be absorbed into the components? This file starts to feel a little bit hectic

ramonjd replied:
Might be a good janitorial follow up to move variation components into `components/global-styles/variations` and then all the related CSS as well?

mtias commented:
This is looking well! I like it doesn't introduce any new APIs. One detail on the placement—I think we should try placing colors within the "Palette" group (not the drilldown) and the type variations within the "Fonts" group. Both cases at the top, with the breakdown right below.

ramonjd commented on packages/edit-site/src/components/global-styles/preview-colors.js:
```diff
@@ -0,0 +1,208 @@
+				>
+					<EditorStyles styles={ editorStyles } />
+					<motion.div
+						style={ {
```
Is this animation block necessary any more for these specific previews?

https://github.com/WordPress/gutenberg/assets/6458278/e769bdff-26f9-4284-b8c7-e7d226394acd

If not, wondering if we can remove it entirely. I was going to, but just wanted to make sure it's doing something that's not obvious.

ramonjd replied:
Sorry, didn't see the transition as the panel opens, as with Typography

https://github.com/WordPress/gutenberg/pull/56622/files#r1503522214

A lot of code though for how much value?

ramonjd commented on packages/edit-site/src/components/global-styles/preview-colors.js:
```diff
@@ -0,0 +1,208 @@
+						} }
+						initial="start"
+						animate={
+							( isHovered || isFocused ) &&
```
If we don't require motion, then I think the `isFocused` prop is also unnecessary (?)

ramonjd commented on packages/edit-site/src/components/global-styles/variations-typography.js:
```diff
@@ -0,0 +1,161 @@
+		? getFamilyPreviewStyle( headingFontFamilies )
+		: {};
+	return (
+		<motion.div
```
Same as color. I wear glasses, but I can't see any animation? Could we ditch motion here?


https://github.com/WordPress/gutenberg/assets/6458278/a7fb5a5f-14d6-4f5a-9557-8bc8a6d46abe

ramonjd replied:
Sorry, I didn't see the transition as the panel opens.

https://github.com/WordPress/gutenberg/assets/6458278/2b3ffaeb-814a-4512-aaf8-1f7fa90db3a5

ramonjd commented on packages/edit-site/src/components/global-styles/variations-typography.js:
```diff
@@ -0,0 +1,161 @@
+	const { base } = useContext( GlobalStylesContext );
+	const uniqueTypographyVariations = [];
+	const uniqueTypographyNames = [];
+	const isDup = ( x, y ) => {
```
If we're using `A` and `a` only, we could probably remove a bit of this code (`uniqueTypographyNames`, `isDup`, `getFontFamilyNames`)  E.g., 

```js
	const uniqueTypographyVariations = Object.values(
		typographyVariations?.reduce( ( acc, variation ) => {
			const [ bodyFontFamily, headingFontFamily ] = getFontFamilies(
				mergeBaseAndUserConfigs( base, variation )
			);
			if (
				bodyFontFamily?.name &&
				headingFontFamily?.name &&
				! acc[
					`${ bodyFontFamily?.name }:${ headingFontFamily?.name }`
				]
			) {
				acc[
					`${ bodyFontFamily?.name }:${ headingFontFamily?.name }`
				] = variation;
			}

			return acc;
		}, {} )
	);
```

ramonjd replied:
I'll commit this - if it breaks I'll revert. YOLO!

ramonjd commented on packages/edit-site/src/components/global-styles/screen-colors.js:
```diff
@@ -29,6 +31,12 @@ function ScreenColors() {
 	const settings = useSettingsForBlockElement( rawSettings );
 
+	const colorVariations = useCurrentMergeThemeStyleVariationsWithUserConfig( {
+		property: 'color',
```
I think I'll work on a variation of this function that accepts an array of properties. 

Here's twentytwentythree getting gradient bleed: 

https://github.com/WordPress/gutenberg/assets/6458278/dca25a59-3c5d-4340-9e68-42d977865718

ramonjd replied:
Actually, withdraw the above comment.

It's because I forgot to reinstate [removePropertyFromObject](https://github.com/WordPress/gutenberg/pull/56622/commits/1d25ce937aee5a9bb538b4d4f9c7dc94c7c745bd#diff-f27afb89c8e1505893fb5dd62b88e81570c0c4032195d2a5b79774a2b59de69bR61).

Sorry!

I'll reinstate.

ramonjd reviewed (commented):
Looking excellent. Thanks for the changes.

I'll push some minor updates if you don't mind, and will also work on sending an array of properties to `useCurrentMergeThemeStyleVariationsWithUserConfig` in a separate PR.

But my main question is around motion. Could we ditch it if there are no animations?

ramonjd committed:
Ensuring the cursor for the color previews is a pointer, which is the same as the typography preview
Reducing the vheight of the color panel to match typography
Some light refactoring

ramonjd committed:
Reinstate removing property from user config before merging with variation.
We do this in order that any user config changes with the same property don't "bleed" into variations. The variation should be "pure".
For example, if the user has defined a background color, but the variation has not, we shouldn't merge the two and allow the user background color to persist.
Rather the variation's color properties should take the place of all color properties.

ramonjd commented on packages/edit-site/src/hooks/use-theme-style-variations/use-theme-style-variations-by-property.js:
```diff
@@ -1,13 +1,76 @@
+ * @param {string} property The property to remove.
+ * @return {Object} The modified object.
+ */
+export function removePropertyFromObject( object, property ) {
```
I deleted this function by accident. So just reverting it and adding tests.

It's required to ensure that variations remain "pure". 

Without this, user-defined properties will persist in any merge. For example, if a user has defined a background color and the selected variation has no background color, the user-defined color will persist. 

This is best demonstrated in the 2023 theme. Notice how the gradient backgrounds are merged with other variations:

https://github.com/WordPress/gutenberg/assets/6458278/72fb8ccc-edc9-493c-aedc-b28d91145c0f

scruffian replied:
Thanks I had noticed the bug but I didn't know the cause :)

ramonjd commented on packages/edit-site/src/components/global-styles/variations-typography.js:
```diff
@@ -0,0 +1,161 @@
+			} }
+		>
+			<span style={ headingPreviewStyle }>A</span>
+			<span style={ bodyPreviewStyle }>a</span>
```
```suggestion
			<span style={ headingPreviewStyle }>{ _x( 'A', 'The letter A' ) }</span>
			<span style={ bodyPreviewStyle }>{ _x( 'a', 'The letter a' ) }</span>
```

ramonjd committed:
i18n for letters

ramonjd committed:
Refactor unique font calculation

scruffian commented:
> I think we should try placing colors within the "Palette" group (not the drilldown) and the type variations within the "Fonts" group. Both cases at the top, with the breakdown right below.

I tried this with @rich but our concern was that it pushed Elements settings down.

scruffian commented:
> But my main question is around motion. Could we ditch it if there are no animations?

I think it depends what the want to do about the duplicated component. My feeling is that we should probably refactor the StylePreview component into smaller pieces that can be reused, so that all three previews share the same code, but this will take a bit of unpicking...

scruffian committed:
move color and type variations higher

scruffian commented:
One idea... 
<img width="287" alt="Screenshot 2024-02-27 at 22 45 36" src="https://github.com/WordPress/gutenberg/assets/275961/7a496adb-506f-497c-b8ea-8c2a74414088">

<img width="288" alt="Screenshot 2024-02-27 at 22 45 45" src="https://github.com/WordPress/gutenberg/assets/275961/8fc5af9b-793f-4a09-9b8d-228c785424ea">

(I pushed this to the PR)

Or should we move the presets under the Palette/Font heading?

ramonjd committed:
Use same box shadow as ColorIndicator so that there's a border around colors
This is so that theme colors that are the same value as the background are more noticeable

ramonjd commented on packages/edit-site/src/components/global-styles/preview-colors.js:
```diff
@@ -0,0 +1,216 @@
+	 */
+	const ratio = ratioState ? ratioState : fallbackRatio;
+
+	const { highlightedColors } = useStylesPreviewColors();
```
While testing I noticed an issue with some themes that had only 2 palette colors, e.g., Freddie.

If those colors match heading or background colors, they're being filtered out:

https://github.com/WordPress/gutenberg/blob/f7b735e9408c10f9600b1a0bcf2c2c0f1e827f80/packages/edit-site/src/components/global-styles/hooks.js#L68-L71

The result is that, although there's a variation, the colors are empty:

<img width="284" alt="Screenshot 2024-02-28 at 12 08 24 pm" src="https://github.com/WordPress/gutenberg/assets/6458278/8ad8e200-2ae0-4d09-8394-6355e44c7416" />


CanvasLoader already gets around it this way:

https://github.com/WordPress/gutenberg/blob/f7b735e9408c10f9600b1a0bcf2c2c0f1e827f80/packages/edit-site/src/components/canvas-loader/index.js#L19-L23

Not sure what's appropriate in `<StylesPreviewColors />`

Maybe something like:

```js
	const { paletteColors, highlightedColors } = useStylesPreviewColors();

	const previewColors = highlightedColors.length
		? highlightedColors
		: paletteColors.slice( 0, 2 );
```

ramonjd replied:
Update: this is already an "feature" of trunk, in that style variation previews won't show them either. 

I think it just looks strange in the preset previews. Maybe good for an immediate follow up.

To replicate the above issue, here's a style variation example. The palette colors are the same as the background and text colors so the variation will show nothing.

```json
{
	"version": 2,
	"settings": {
	"color": {
		"palette": [
			{
				"color": "#000000",
				"name": "Base",
				"slug": "base"
			},
			{
				"color": "#ffffff",
				"name": "Contrast",
				"slug": "contrast"
			}
		]
	}
	},
	"styles": {
		"color": {
			"text": "#000000",
			"background": "var(--wp--preset--color--contrast)"
		}
	}
}

```

ramonjd commented on packages/edit-site/src/components/global-styles/preview-colors.js:
```diff
@@ -0,0 +1,216 @@
+											key={ slug }
+											style={ {
+												boxShadow:
+													'inset 0 0 0 1px #0003',
```
I added this to ensure the colors have a border like the ColorIndicator component. 

In some themes, the background color matches the highlights, or are close in contrast:

### Before
<img width="282" alt="Screenshot 2024-02-28 at 12 11 30 pm" src="https://github.com/WordPress/gutenberg/assets/6458278/2e38912d-6af9-4dc0-bdde-6894857ee5e4">


### After

<img width="281" alt="Screenshot 2024-02-28 at 12 11 02 pm" src="https://github.com/WordPress/gutenberg/assets/6458278/6dccd5ba-34eb-4eb0-bf59-9f2c5c7fc725">

richtabor replied:
> I added this to ensure the colors have a border like the ColorIndicator component.


I don't think this is necessary, as I want to update these colors to be more representative of the style by using the text and button colors for the indicators. I will open another issue on that (to do as a follow-up).

ramonjd reviewed (approved):
This is working pretty well, thanks a lot!

I'm happy if @mtias and @richtabor are happy. 😄 

I've tested in about 12 block themes and didn't notice any huge blockers.

### For follow ups

It would be good to have some fallbacks for [highlightedColors](https://github.com/WordPress/gutenberg/blob/f7b735e9408c10f9600b1a0bcf2c2c0f1e827f80/packages/edit-site/src/components/global-styles/preview-colors.js#L104-L104) in case the background/text styles are the same as the available palette colors. See comment: https://github.com/WordPress/gutenberg/pull/56622#discussion_r1505219705

From this [comment](https://github.com/WordPress/gutenberg/pull/56622#discussion_r1501797756): Might also be a good janitorial follow up to move variation components into components/global-styles/variations and then all the related CSS as well

> My feeling is that we should probably refactor the StylePreview component into smaller pieces that can be reused, so that all three previews share the same code, but this will take a bit of unpicking...

Agree. This would be a good refactor follow up.

richtabor commented:
> One detail on the placement—I think we should try placing colors within the "Palette" group (not the drilldown) and the type variations within the "Fonts" group. Both cases at the top, with the breakdown right below.

I agree, hierarchy wise they're a bit in the mix—rather than appropriately placed. 

I think this could work for this initial effort, then we can revisit secondary affordances for when there are more than 6 (perhaps a "View more/all" and a flyout, like the Colors panel):

![CleanShot 2024-02-28 at 10 50 04](https://github.com/WordPress/gutenberg/assets/1813435/d52f3dac-8300-4169-b0a3-de4da764fb74)

scruffian commented:
Design updated:

<img width="297" alt="Screenshot 2024-02-28 at 15 55 39" src="https://github.com/WordPress/gutenberg/assets/275961/b239e121-dec0-479d-a5c2-1915f37b4bab">
<img width="298" alt="Screenshot 2024-02-28 at 15 55 31" src="https://github.com/WordPress/gutenberg/assets/275961/a2e0e826-f28d-4bc1-b90e-17cd30929e7e">

scruffian commented:
I'm going to bring this in, but just noting the followups here:

1. Consider breaking the StylePreveiw component into smaller components that can be shared across these different instances, and clean up the extra CSS (https://github.com/WordPress/gutenberg/pull/56622/files#r1501797756)
2. Reconsider the colors used in the StylePreview
3. Deal with the case where presets are appearing blank (https://github.com/WordPress/gutenberg/pull/56622#discussion_r1505232709)

scruffian commented:
Sorry I did the merge and forgot to add the props to it again :(
```
Co-authored-by: scruffian <scruffian@git.wordpress.org>
Co-authored-by: ramonjd <ramonopoly@git.wordpress.org>
Co-authored-by: richtabor <richtabor@git.wordpress.org>
Co-authored-by: mtias <matveb@git.wordpress.org>
Co-authored-by: MaggieCabrera <onemaggie@git.wordpress.org>
```

richtabor commented:
> Reconsider the colors used in the StylePreview

Follow up issue here: https://github.com/WordPress/gutenberg/issues/59444

richtabor commented:
Good stuff! 

The best part of this effort is that all existing block themes will inherit this without having to do a thing. Anyone running a block theme can use any combination of colors or fonts, from any of their theme variations. Very nice.

matiasbenedetto commented:
Hi :wave: , I'm sorry to be late to this conversation. The typography presets are not working as users would expect. Its behavior feels buggy and unpredictable. I submitted an issue around this: https://github.com/WordPress/gutenberg/issues/59574

timnicholson commented:
I've noticed that color palettes don't show up as such if the colors-only json in the /styles directory has a "style" section in it at all. Even if the syles section only has a CSS value. I have some light CSS that is needed to change the font colors on certain background colors. 

For example, this doesn't show up as a color palette in v6.6. I assume the reasoning is that if a global style contains anything other than color settings, it should be retained as a global style and not generate a preset color palette that really limits what you can do to default readable text based on the background color. 
```
{
	"$schema": "https://schemas.wp.org/trunk/theme.json",
	"version": 3,
	"title": "Midnight Blue, Blue, Yellow",
	"styles": {
		"css": ".has-background.has-tertiary-background-color { color: var(--wp--preset--color--foreground); }"
	},	
	"settings": {
		"color": {
			"palette": [
				{
					"slug": "primary",
					"color": "#34495e",
					"name": "Primary (Midnight Blue)"
				},
				{
					"slug": "secondary",
					"color": "#3498db",
					"name": "Secondary (Blue)"
				},
				{
					"slug": "tertiary",
					"color": "#f1c40f",
					"name": "Tertiary (Yellow)"
				}
```
Any chance that "color pairs" are in the works for theme.json and global styles? Meaning the ability to specify a background and text color at the same time? Something like this:
```
{
	"slug": "primary",
	"color": "#34495e",
	"text": "#ffffff",
	"name": "Primary (Midnight Blue)"
},
```

scruffian commented:
@timnicholson your assumption is correct. If the style variation contains any keys other than color it will be treated as a full style variation, not a color palette. I would be reluctant to make exceptions for specific keys because where does that end? I think the better option is to fix the issues that the CSS is trying to solve by improving Global Styles.

> Any chance that "color pairs" are in the works for theme.json and global styles?

Not that I know of it.

aicuaboom commented:
👿

timnicholson commented:
I just noticed that WordPress v6.6 is now outputting CSS that sets the text color to the foreground color on colored backgrounds. And it uses specificity with `.has-background` as well, which overrides my theme CSS that uses `.has-primary-background-color`. I thought 6.6 was supposed to reduce CSS specificity :-/

   ` .has-background.has-primary-background-color {
        color: var(--wp--preset--color--foreground)
    }`

I was able to fix that with the additional specificity. 

But another issue is the two little color dots on the global theme styles are all gray no matter what the color scheme is. Does anyone know what colors it is actually pulling for those? They used to pull the first and second color in the palette (so I made the first two the "interesting" ones), but now they are all just gray.

richtabor commented:
> I just noticed that WordPress v6.6 is now outputting CSS that sets the text color to the foreground color on colored backgrounds. 

Do you mind opening a new issue with more details of this (visuals, active theme, etc). It's not something I am aware of. 

> But another issue is the two little color dots on the global theme styles are all gray no matter what the color scheme is. 

They're using background color, heading and button/link color.

timnicholson commented:
> > I just noticed that WordPress v6.6 is now outputting CSS that sets the text color to the foreground color on colored backgrounds.
> 
> Do you mind opening a new issue with more details of this (visuals, active theme, etc). It's not something I am aware of.
> 
Sorry, I jumped the gun on this. I thought it was WordPress outputting CSS for the foreground color, but it wasn't. The "bug" is actually that the Global Theme Styles preview doesn't swap out the "additional CSS" from the theme.json partials. So when I chose a style that used dark text on the light primary color, that CSS "stuck" no matter what other style I chose. 

> > But another issue is the two little color dots on the global theme styles are all gray no matter what the color scheme is.
> 
> They're using background color, heading and button/link color.
I can't get those two dots to be any color other than the foreground color. Even simply using the Block Editor Settings to change the button or link color doesn't change the color of the two dots. The only way I can get those dots to change is if I put something truly unique into the foreground color setting, like to include comments like this. This seems to override some WordPress logic that is checking to see if the foreground color in the global style json is different than that of the main theme.json. By including these comments, it wont' equal "#555555" in theme.json and thus will display the button color in both dots. Very odd bug. 
```
	"styles": {
		"color": {
			"background": "var(--wp--preset--color--base)",
			"text": "#555555 /*var(--wp--preset--color--contrast)*/"
		},
```
